enterprise/internal/batches/service/ui_publication_states.go

```diff
@@ -0,0 +1,103 @@
+package service
+
+import (
+	"context"
+
+	"github.com/cockroachdb/errors"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/batches/store"
+	btypes "github.com/sourcegraph/sourcegraph/enterprise/internal/batches/types"
+	"github.com/sourcegraph/sourcegraph/lib/batches"
+)
+
+// UiPublicationStates takes the publicationStates input from the
+// applyBatchChange mutation, and applies the required validation and processing
+// logic.
+//
+// External users must call Add() to add changeset specs to the struct, then
+// process() must be called before publication states can be retrieved using
+// get().
+type UiPublicationStates struct {
+	rand map[string]batches.PublishedValue
+	id   map[int64]*btypes.ChangesetUiPublicationState
+}
+
+// Add adds a changeset spec to the publication states.
+func (ps *UiPublicationStates) Add(rand string, value batches.PublishedValue) error {
+	if ps.rand == nil {
+		ps.rand = map[string]batches.PublishedValue{rand: value}
+		return nil
+	}
+
+	if _, ok := ps.rand[rand]; ok {
+		return errors.Newf("duplicate changeset spec: %s", rand)
+	}
+
+	ps.rand[rand] = value
+	return nil
+}
+
+func (ps *UiPublicationStates) get(id int64) *btypes.ChangesetUiPublicationState {
+	if ps.id != nil {
+		return ps.id[id]
+	}
+	return nil
+}
+
+type ListChangesetSpeccer interface {
+	ListChangesetSpecs(context.Context, store.ListChangesetSpecsOpts) (btypes.ChangesetSpecs, int64, error)
+}
+
+var _ ListChangesetSpeccer = &store.Store{}
+
+// prepare looks up the random changeset spec IDs, and ensures that the
+// changeset specs are attached to the batch spec and are eligible for a UI
+// publication state.
+func (ps *UiPublicationStates) prepare(ctx context.Context, s ListChangesetSpeccer, batchSpecID int64) error {
```

Comment:
nit: I think that is a bit more precise? 

```suggestion
func (ps *UiPublicationStates) prepareAndValidate(ctx context.Context, s ListChangesetSpeccer, batchSpecID int64) error {
```

Reply:
Done (with the appropriate other changes).